src/components/Table/GridTable.tsx

```diff
@@ -80,6 +81,13 @@ export type GridSortConfig =
       onSort: (orderBy: string | undefined, direction: Direction | undefined) => void;
     };
 
+/** Allows listening to per-kind row selection changes. */
+export type OnRowSelect<R extends Kinded> = {
+  [K in R["kind"]]?: DiscriminateUnion<R, "kind", K> extends { data: infer D }
+    ? (data: D, isSelected: boolean, opts: { row: GridRowKind<R, K>; api: GridTableApi<R> }) => void
+    : (data: undefined, isSelected: boolean, opts: { row: GridRowKind<R, K>; api: GridTableApi<R> }) => void;
+};
```

Comment:
**question:** Does this force the user to add a callback function per kind? Or can they decide which kinds to add callbacks to? 
I'm wondering what will happen if a user selects a parent row, which in turn selects 10 child rows. Are we going to have 11 callbacks executed - one on the parent and 10 on child rows? It looks that way to me.
Would it be more useful and performant to execute a single callback that maybe includes all selection changes? Maybe an array of tuples `[isSelected, row][]`. Though, I also wonder if it's just worth returning an array of the selected rows vs rows that were deselected.
I'm thinking of use cases where we currently update a form state of selected items (granted this should be a multi-select component, not a table), or when we are enabling/disabling bulk actions based on table selections. Seems like this could be a bit slow to execute a single callback at a time.

Reply:
> Does this force the user to add a callback function per kind? Or can they decide which kinds
> to add callbacks to?

Yep, they can add only what they're interested in; the `?` on the `]]?: ...` line makes each kind key optional.

> Are we going to have 11 callbacks executed - one on the parent and 10 on child rows?

Yes.

> Though, I also wonder if it's just worth returning an array of the selected rows vs
> rows that were deselected.

I think it depends on what the user is doing. If the user just wants "an array of all selected rows", they can use the existing `api.getSelectedRows`.

The code I'm refactoring was doing "when _this_ row changes, clear fields x/y/z on it", so, I dunno, I think the api of "this row changed, do what you want with it" is more intuitive for that use case...

> when we are enabling/disabling bulk actions based on table selections.

Maybe this is why we're talking past each other on "row.onSelect" vs. "props.onSelect" change--I think you're thinking of more bulk/table-level reactions, like "_all_ of my selected rows are now x/y/z, so do this", where-as I'm refactoring this code:

https://github.com/homebound-team/internal-frontend/blob/main/src/routes/projects/invoices/invoice-v2/invoice-steps/SelectFixedItems.tsx#L104

Which is very this-row-focused/scoped in terms of the reaction it wants to drive.

Maybe that is rationale for having "two APIs for the same thing?" We have `row.onSelect` for "I just want to mutate things in this row when it changes state" (without having to deduce that from watching "the full list of rows is now x/y/z"), and we have `props.onSelectionChange` when you want to react to "the entire list of selected rows is now x/y/z".

If so, yeah, we could simplify `onSelectionChange` to just be "here's the full list of rows", which per prior comment is really the same thing as "run a `useComputed`/`useReaction` against the `api.getSelectedRows` observable", but admittedly with a more "react-ish" API, i.e. just give us a callback to call, instead of having to watch the api.

I'm good with that, if you are.

Reply:
If you want to double-down on "we shouldn't have any `GridDataRow` callbacks i.e. no row.onSelect", maybe we're talking about two separate props, i.e.:

* `GridTableProps.onRowSelect` --> fired on individual row is/is not selected (this is the API I want for the refactoring at hand), and
* `GridTableProps.onSelectionChange` --> fired with the complete list of new rows, regardless of kind, which I believe is what you're expecting/asking for

Reply:
I see what you're referring to now regarding the `SelectFixedItems` component.
That does seem nice to have the row specific actions on, so I see how we'd benefit there.
To me, there seems to be a bit of a code-smell when having callbacks on a GridDataRow. So I'd be okay with your last suggestion. Then we'd definitely want the `onRowSelect` to return `(isSelected: boolean; row: GridDataRow<R>)` so the user could access the row's data/formState when necessary.
But the code-smell isn't strong enough to absolutely push me that route. I'd be okay with us having it per row. Will be curious to see if that ever becomes a problem or not, though.

Reply:
Cool, pushed up the removal of `GridDataRow.onSelect`, so we have just `GridTableProps.onRowSelect`. We can follow up with a `GridTableProps.onSelectionChange`, which gives the user the whole list, at some point.